### ISSUE TITLE
The `extern "C"` ABI of SIMD vector types depends on target features (tracking issue for `abi_unsupported_vector_types` future-incompatibility lint)

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#116558]
The following program has UB, for very surprising reasons:
```rust
use std::mem::transmute;
#[cfg(target_arch = "x86")]
use std::arch::x86::*;
#[cfg(target_arch = "x86_64")]
use std::arch::x86_64::*;

extern "C" fn no_target_feature(_dummy: f32, x: __m256) {
    let val = unsafe { transmute::<_, [u32; 8]>(x) };
    dbg!(val);
}

#[target_feature(enable = "avx")]
unsafe fn with_target_feature(x: __m256) {
  // Here we're seemingly just calling a safe function... but actually this call is UB!
  // Caller and callee do not agree on the ABI; specifically they disagree on how the
  // `__m256` argument gets passed.
  no_target_feature(0.0, x);
}

fn main() {
    assert!(is_x86_feature_detected!("avx"));
    // SAFETY: we checked that the `avx` feature is present.
    unsafe {
        with_target_feature(transmute([1; 8]));
    }
}
```
The reason is that the ABI of the `__m256` type depends on the set of target features, so the caller (`with_target_feature`) and callee (`no_target_feature`) do not agree on how the argument should be passed. The result is a [vector half-filled with junk](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ee15f570ee5df1ecbec6321c44066008). (The same issue also arises in the other direction, where the caller has fewer features than the callee: [example here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=78a91a7266b7e39fe539c1e8d8c9892d).)

Currently, we have no good way to do correct code generation here. See https://github.com/rust-lang/rust/issues/132865 for a discussion of how we could support such code in the future; it will require some non-trivial work. So instead, the current plan is to reject such code entirely.

This is the tracking issue for the lint that moves us in that direction. The hope is that passing SIMD vectors across a C ABI is sufficiently rare, and most of the cases being rejected have anyway already been broken, that we can reject this without much of an ecosystem impact. Crater showed no regressions. 

This is linted against since Rust 1.84, and shown in future breakage reports starting with Rust 1.85. The plan is to move to a hard error with Rust 1.87. 

**When this becomes a hard error:**

- We can probably remove the error tracked at https://github.com/rust-lang/rust/issues/63068?
- And also the warning triggered [in this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10c712ef153066a8c1829f6c7f41a3f4)?

----

**Original issue text:**

I'm not sure if this is currently even properly documented? We are mentioning it in https://github.com/rust-lang/rust/pull/115476 but the program above doesn't involve any function pointers so we really cannot expect people to be aware of that part of the docs. We show a general warning about the type not being FFI-compatible, but that warning shows up a lot and anyway in this case both caller and callee are Rust functions!

I think we need to do better here, but backwards compatibility might make that hard. @chorman0773 suggested we should just reject functions like `no_target_feature` that take an AVX type by-val without having declared the AVX feature. That seems reasonable; a crater run would be needed to assess whether it breaks too much code. An alternative might be to have a deny-by-default lint that very clearly explains what is happening.

There are also details to work out wrt what exactly the lint should check. Newtypes around `__m256` will have the same problem. What about other larger types that contain `__m256`? Behind a ptr indirection it's obviously fine, but what about `(__m256, __m256)`? If we apply the `ScalarPair` optimization this will be passed in registers even on x86.

A possible place to put the check could be [somewhere around here](https://github.com/rust-lang/rust/blob/8fa7bdf191d8658e5ab670766ac0b2d40c7d83c2/compiler/rustc_codegen_ssa/src/codegen_attrs.rs#L279-L309).

In terms of process, I am not sure if an RFC is required; a t-compiler MCP might be sufficient. Currently we accept code that clearly doesn't do what it looks like it should do.

And finally -- are there any other targets (besides x86 and x86-64) that have target-features that affect the ABI? They should get the same treatment.

Note that this is different from https://github.com/rust-lang/rust/issues/116344 in two ways:
- This issue only affects non-"Rust" ABIs; the other issue affects even the "Rust" ABI. (The Rust ABI passes `__m256` by-pointer exactly to work-around this issue.)
- This issue comes about when the user enables extra features (which is possible also locally via `#[target_feature]`), the other issue comes about when the user *disables* features (which is only possible on a per-crate level via `-C`, and if you're mixing crates with different `-C` flags then you're already already on very shaky grounds -- we do that with std but nobody else really gets to do that, I think).

Cc @workingjubilee more ABI fun ;)

**Comment by chorman0773** [comment]
For a hard error, I think it probably should go through RFC. I can start drafting that now. As for what it would check, I'd expect a hard error on any time a simd type is passed by value and the appropriate feature is unavailable, even if the specific ABI wouldn't necessarily pass it in registers, including inside a structure, and the lint should probably check the same thing.

**Comment by RalfJung** [comment]
That can have potentially surprising effects, e.g. when a closure captures something of type `__m256` and the closure gets passed by-value to a generic function.

It also means adding adding a private field that contains by-val `repr(simd)` data can be a semver-breaking change. Cc @ehuss

**Comment by coastalwhite** [comment]
With RISC-V, you have `Zfinx` and `Zdinx` which have hard-floats but have moved the floats to the normal general purpose registers. In this case, forbidding the passing of floats by-value, if the `f` (RISC-V's version of single-precision float) target is not present, is very restrictive and probably not what you would want. Otherwise, allowing the passing-by-value of floats between `f` and `Zfinx` is also UB.

**Comment by RalfJung** [comment]
But the ABI only changes when `f` is present or not present, I assume? When `f` is present, `Zfinx` just means more instructions can be used but the ABI still says `f32` is passed on float registers? And when `f` is not present, softfloat and `Zfinx` use the same ABI?

So that sounds like whether or not `f` is present should be a different target triple. Code with vs without `f` simply cannot be linked together. But having or not having `Zfinx` doesn't make an ABI difference. Or am I misunderstanding?

**Comment by coastalwhite** [comment]
No, that is exactly it. Do I understand correctly that you essentially want to do something similar to `thumb` with the `eabi` and `eabihf`?

**Comment by RalfJung** [comment]
Yeah for the Rust ABI we do apply a work-around (it always passes such types in memory), so that call is fine.

(I updated the example to actually use the Rust ABI for `with_target_feature`.)

**Comment by Skgland** [comment]
Would adding a [`inline(never)` Rust ABI intermediate function](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=169ffff9fbc948c0552e724dc99d8eac) be a valid workaround to prevent UB here?
[Unconditional code generation #[target_feature]: Note 1](https://rust-lang.github.io/rfcs/2045-target-feature.html#unconditional-code-generation-target_feature) making `#[inline(never)]` required as otherwise the compiler may inline the function and then we would be back to the incorrect feature set.

**Comment by RalfJung** [comment]
Inlining should be aware of target features, so `inline(never)` should not be needed. If it is needed that's a critical bug in the inliner that we should track and fix.

Yes such an intermediate Rust ABI function works around the problem successfully. But the fundamental problem remains with function pointers: we have to decide whether an `extern "C" fn(__m256)` has the AVX ABI or some fallback ABI, without having any clue where and how the function it points to was declared. We probably want it to have the AVX ABI since people that use AVX and interface between Rust and C code likely will have their C code built in a way that it uses the AVX ABI. This means all `extern "C" fn` declared in Rust that take `__m256` as arguments should do this the AVX style, no matter the set of enabled target features -- or they should fail to build.

**Comment by RalfJung** [comment]
Ah, the `inline` actually *does* make a difference here... well that's bad, I opened https://github.com/rust-lang/rust/issues/116573 to track that.

**Comment by briansmith** [comment]
The compiler does report `improper_ctypes_definitions` warnings for the given code so I guess it isn't unexpected? It would e better to have an example that doesn't trigger `improper_ctypes_definitions`.

**Comment by workingjubilee** [comment]
> The compiler does report improper_ctypes_definitions warnings for the given code so I guess it isn't unexpected? It would e better to have an example that doesn't trigger improper_ctypes_definitions.

And that's supposed to be just a _warning_. We should hard error if we know everything is fucked and there's no salvaging the compilation.

**Comment by RalfJung** [comment]
There's no C code involved here so I would argue it is reasonable to silence the lint for the example. The lint talks about FFI, but the example does no FFI, this is all Rust-to-Rust calls.

**Comment by briansmith** [comment]
Is there an example that can trigger this with only FFI-safe types though?

**Comment by chorman0773** [comment]
I've actually suppressed that lint for `__m128` on sse3 only code (so xmm registers 100% exist) that was experiencing a perf regression when the function used default ABI.
But also I wouldn't trust `improper_ctypes{,_definitions}` to not get blanket suppressed in FFI heavy code, so at the very least a lint for this should be it's own thing. improper_ctypes is a clippy lint in rustc.

**Comment by tmandry** [comment]
I don't think `#[target_feature]` should ever affect the ABI of function pointers called from within them, especially not `extern "C"`, ones where the C ABI is defined statically at the TU level.

If we did anything with ABI based on `#[target_feature]`, it would have to be encoded in the type of the function pointers somehow. So for example, an `extern "C"` function with an avx calling convention would really be `extern "C+avx"` or similar.

**Comment by RalfJung** [comment]
@tmandry yeah agreed. Sadly target_feature 1.0 was stabilized in a form where `#[target_feature]` does affect ABI, and `core::arch` SIMD types where stabilized in a form where `-C target-feature` affects ABI.

**Comment by RalfJung** [comment]
@briansmith our ctypes lints seem to have enough gaps to be easily circumvented; [here's a reproducer](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=6b82f949a5adda17519355792f83332e) that the lint doesn't catch. It's using a closure to pass the value -- closures that capture a single field are newtypes around that field so they get the field's ABI, but the improper-ctypes lint ignores closures.

**Comment by workingjubilee** [comment]
More to the point, there's no formal definition of "FFI-safe" and "FFI-unsafe" in the Rust language. As far as I can tell, the actual thing that bears the obligation to be ABI-correct is when you do

```rust
extern "C" {
    fn c_func(arg: SomeType) -> ReturnType;
}

fn rust_func() -> ReturnType {
    let arg = SomeType::random();
    // here, the Rust caller actually discharge the obligation...
    // which is honestly slightly unfortunate:
    // in real scenarios, we might not be the author of the `extern "C"` block!
    unsafe { c_func(arg) }
}
```

**Comment by RalfJung** [comment]
Do we know if clang is doing anything clever here? I assume C has exactly the same problem, since their function types don't track target features either.

**Comment by chorman0773** [comment]
clang does not fix this issue. See [llvm-project#64706](https://github.com/llvm/llvm-project/issues/64706)

**Comment by RalfJung** [comment]
That's a different issue (mismatch with GCC). This here is about mismatch within code compiled by the same compiler, but with different locally set target features.

**Comment by chorman0773** [comment]
Ah right. It does, in fact, mismatch with itself without the attribute.

**Comment by RalfJung** [comment]
Do you happen to have a godbolt link demonstrating that? :)

**Comment by chorman0773** [comment]
https://godbolt.org/z/ae8aTYa9d

Interpreting the assembly: In the global (`-mavx`) case, it passes `__m256` in `ymm` registers, and returns them in `ymm0`. For the inactive case, it passes on the stack and returns in `xmm0:xmm1`. For the locally-enabled only case, it passes on the stack, and returns in `ymm0` (according to Sys-V, there shouldn't be a difference with the globally enabled case).

**Comment by RalfJung** [comment]
Turns out that clang is able to [detect and warn](https://godbolt.org/z/x8f4bMfPj) at least some of these cases. Clang also allows putting `__attribute__((target("avx")))` annotations at function declarations and uses them to decide whether or not to warn/error. In contrast, we allow `#[target_feature]` only on definitions, not on imports (`extern "C" {}` blocks).

One way to resolve this problem would be a new post-monomorphization check: traverse the argument (and return) types of the function being built (if it's a non-Rust ABI), as well as the argument and return types of every non-Rust-ABI function signature we call. If we find any `repr(simd)` types in there that's not behind a pointer indirection, ensure that vectors of that size are supported by the available target features. I don't know if there's any good cross-target way to do this; if not, we'll need a big match on the target architecture and hard-code which features need to be enabled for which vector size.

This is stricter than what clang does, but it's also safer, so if we can get away with it I think this would be my preferred approach. We don't have a good central place for post-mono checks, so we'll need to figure out what is the best way to share this between the codegen backends and Miri. Also, such a check would likely be optimization-dependent, since if an ABI-incompatible function call gets optimized away we wouldn't see it post-mono. We have to do the check post-mono though since we support generic `extern "C"` fn and there is no trait for "does not contain a SIMD type by-value" (and it would be a breaking change to start requiring such a trait).

**Comment by chorman0773** [comment]
Yep (in my case, right down at the codegen level - codegen-x86 would reject the function when finding parameter locations). I actually have an RFC-being-written for prescribing this behaviour at the rust level.

**Comment by RalfJung** [comment]
> I actually have an RFC-being-written for prescribing this behaviour at the rust level.

Great!

Given that we currently do rather nonsensical things at the ABI level, I was going to suggest we should just implement a band-aid without waiting for an RFC. But that doesn't preclude designing something more proper. :)

**Comment by chorman0773** [comment]
One note is this text that I put
> When a function is called via an fn-item type (not a function pointer type), the target features of the callee are considered. [...]
> Implementations of Rust would be required to ensure this call is performed properly (with parameters passed/returned in the appropriate registers), which may entail the use of a shim function or alternative code generation techniques. 

This may require some extra work with llvm, or on the rustc side.

**Comment by RalfJung** [comment]
Interesting, you seem to plan to accept more code than I would. Anyway we can discuss the RFC once it is ready. :)

**Comment by chorman0773** [comment]
(Actually, I could probably open it in a less formal capacity - I think there is some pre-discussion to be had before filing a full RFC)

**Comment by RalfJung** [comment]
There's now [a pre-RFC](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Pre-RFC.20discussion.3A.20Forbidding.20SIMD.20types.20w.2Fo.20features)  by @chorman0773  that proposes a way to resolve this problem.

**Comment by RalfJung** [comment]
To my great surprise, we knew about this issue since 2019: https://github.com/rust-lang/rust/issues/64609. We just did nothing about it for >4 years...

**Comment by RalfJung** [comment]
This can replace several old issues marked I-unsound, so let's mark this one as well.

**Comment by ehuss** [comment]
I'm not sure if people are aware, but `compiler_builtins@0.1.146` is currently triggering this future-incompatibility warning. This is showing up on the `test-various` and `dist-various-2` CI jobs with x86_64-unknown-uefi. This can be reproduced (on x86_64) with `./x build library/std --stage=1 --target x86_64-unknown-uefi` or in the `compiler-builtins` repo, running `cargo b --target x86_64-unknown-uefi` (at commit 3c09866aa4c699ed5f430567f6de97ed56b65126).

The warnings look something like:

```
The package `compiler_builtins v0.1.146` currently triggers the following future incompatibility lints:
> warning: this function definition uses a SIMD vector type that (with the chosen ABI) requires the `sse` target feature, which is not enabled
>    --> /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/compiler_builtins-0.1.146/src/macros.rs:252:9
>     |
> 71  | /  macro_rules! intrinsics {
> 72  | |      () => ();
> ...   |
> 252 | |          pub extern $abi fn $name( $($argname: $ty),* ) $(-> $ret)? {
>     | |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function defined here
> ...   |
> 426 | |          intrinsics!($($rest)*);
>     | |          ----------------------
>     | |          |
>     | |          in this macro invocation (#2)
>     | |          in this macro invocation (#3)
> ...   |
> 577 | |      );
> 578 | |  }
>     | |  -
>     | |  |
>     | |  in this expansion of `intrinsics!` (#1)
>     | |__in this expansion of `intrinsics!` (#2)
>     |    in this expansion of `intrinsics!` (#3)
>     |
>    ::: /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/compiler_builtins-0.1.146/src/float/conv.rs:395:1
>     |
> 395 |  / intrinsics! {
> 396 |  |     #[arm_aeabi_alias = __aeabi_f2uiz]
> 397 |  |     pub extern "C" fn __fixunssfsi(f: f32) -> u32 {
> 398 |  |         float_to_unsigned_int(f)
> ...    |
> 443 |  | }
>     |  |_- in this macro invocation (#1)
>     |
>     = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
>     = note: for more information, see issue #116558 <https://github.com/rust-lang/rust/issues/116558>
>     = help: consider enabling it globally (`-C target-feature=+sse`) or locally (`#[target_feature(enable="sse")]`)
```

**Comment by RalfJung** [comment]
Oh, good point... I was about to make this a hard error, I guess then we would have noticed. ;)

Cc @Amanieu looks like something is wrong with one of the intrinsics? Sadly we don't get a great span as there are some macros involved.

**Comment by RalfJung** [comment]
Hm, but there are no SIMD types in the signature for any of these intrinsics. Very strange.

**Comment by RalfJung** [comment]
Some of these errors are about `U64x2`, but the vast majority is about `u128`. This is probably caused by the following ABI adjustment:

https://github.com/rust-lang/rust/blob/1f37b9a643c28858cc3fac392483bf3f541182e2/compiler/rustc_target/src/callconv/x86_win64.rs#L26-L30

This was added in https://github.com/rust-lang/rust/pull/134290, Cc @tgross35. Surely on a target that does not have float or SIMD registers, i128 does not get passed in `xmm0`?

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/137094 fixes most of the warnings. The remaining ones seem to be about U64x2 which indeed does not make sense as a type on a softfloat target; I created a separate issue for that: https://github.com/rust-lang/compiler-builtins/issues/758.

**Comment by oconnor663** [comment]
@RalfJung I've noticed new warnings as a result of this change. Here's a CI run: https://github.com/BLAKE3-team/BLAKE3/actions/runs/15326026728/job/43120654065. Reproducing that locally:

```
$ git clone https://github.com/BLAKE3-team/BLAKE3                                                                                                  
...
$ cd BLAKE3                                                                                                                                        
$ cross rustc --release --target i686-unknown-linux-musl -- -C target-cpu=i386                                                                     
...
warning: this function definition uses SIMD vector type `std::arch::x86::__m128i` which (with the chosen ABI) requires the `sse` target feature, wh
ich is not enabled                                                                                                                                 
  --> src/rust_sse2.rs:75:1                                                                                                                        
   |                                                                                                                                               
75 | / unsafe fn g1(                                                                                                                               
76 | |     row0: &mut __m128i,                                                                                                                     
77 | |     row1: &mut __m128i,                                                                                                                     
78 | |     row2: &mut __m128i,                                                                                                                     
79 | |     row3: &mut __m128i,                                                                                                                     
80 | |     m: __m128i,                                                                                                                             
81 | | ) {                                                                                                                                         
   | |_^ function defined here
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #116558 <https://github.com/rust-lang/rust/issues/116558>
   = help: consider enabling it globally (`-C target-feature=+sse`) or locally (`#[target_feature(enable="sse")]`)
...
```

The code in [that file](https://github.com/BLAKE3-team/BLAKE3/blob/60ed9295f58ba975df58a67b2bfe435f478d05ea/src/rust_sse2.rs) is divided into two types of functions, public ones annotated with `#[target_feature(enable = "sse2")]`, and private ones annotated with `#[inline(always)]`. I had thought (or read somewhere) that `inline(always)` functions assumed the CPU features of their caller and didn't need to be annotated. Either way it seems to be incompatible with `target_feature`. What's should this code be doing differently?

**Comment by RalfJung** [comment]
@oconnor663 please file a new issue.

I'm surprised this only came up now, this was an FCW for 4 releases...

**Comment by oconnor663** [comment]
Weird, I'm pretty sure this test target has been running regularly for years. Thanks for getting back to me so quickly. Opened: https://github.com/rust-lang/rust/issues/141848

**Comment by RalfJung** [comment]
Well, FCW are warnings, so if nobody looks a the CI output and CI does not set "-D warnings" and people ignore the warnings in local builds they can get missed.

**Comment by RalfJung** [comment]
Ah this is with 1.87, not with nightly, I see -- so it is still a warning. It'll be a hard error with 1.88.

But it has been a warning for local crates since 1.84, and for all dependencies since 1.85.
